[CalyxToFSM] Initial Calyx-to-FSM control schedule lowering

mortbopet opened the pull request:
This commit represents a lowering pass for converting a Calyx control schedule (seq/if/while/enable) into an FSM representation. The FSM is embedded within the Calyx component, and references both the group symbols and cell SSA values.

The lowering method is a fairly straight forward conversion, which leaves plenty of canonicalization opportunities to remove redundant states in the generated FSM. However, doing it as presented in this PR lends itself to some clean code (which I prefer, rather than prematurely optimizing during lowering) as well as meaningful naming of states (which i think is fairly important).

Marked as draft due to mixing in changes to the FSM dialect which needs to be discussed. Once things are settled, FSM dialect and FSMGraph changes will be factored out into a separate PR.

Things to be discussed:
- MachineOp is **no more** `IsolatedFromAbove`; this is a requirement for the FSM to be able to reference the symbols and SSA values defined within the component. If we find that the presented style of lowering is compelling, but we still want an `IsolatedFromAbove` FSM machine, one option could be to introduced a new `EmbeddedMachineOp` to allow for this kind of behaviour.
- ~~I added a `fsm.ref` (not sure about the name) to represent _a reference to some symbol_. The semantics of a reference is implementation defined, and merely serves to convey the semantics of associating a symbol reference with a given state. `fsm.ref` is call-like and can provide arguments to references, as well as return values.~~
  - Edit: Removed in favor of relaxing `calyx.enable` verifier.

Graph print of the `lower-while.mlir` test:
<img src="https://user-images.githubusercontent.com/16338943/168989336-c9a59b75-c62f-432c-8529-f77617121a57.png" alt="image" width="500" class="center"/>

mikeurbach commented on include/circt/Dialect/FSM/FSMOps.td:
```diff
@@ -113,6 +114,24 @@ def TriggerOp : FSMOp<"trigger", []> {
+  let summary = "Reference some side effect through a symbol reference";
+  let description = [{
+    `fsm.ref` represents a reference to some symbol. The semantics of
+    a reference is implementation defined, and merely serves to convey
```
I don't know if saying the semantics is implementation defined is really saying anything. Can you give a more specific semantics to what this op is, may or may not be used for, and why it would be used in the IR?

mikeurbach commented on test/Conversion/CalyxToFSM/lower-while.mlir:
```diff
@@ -0,0 +1,130 @@
+// CHECK:       }
+// CHECK:     }
+// CHECK:     fsm.state "seq_1_while_if_else_seq_0_do_add" output {
+// CHECK:       fsm.ref @do_add() : () -> ()
```
In the sketch you showed before, this literally embedded a calyx.enable, but now it is this fsm.ref op. If this approach is going to be scalable, I think you will need to more formally define what this fsm.ref means.

Knowing a bit about what you are going for, it seems like this is a higher-level concept that will later be transformed into a lower-level concept. My understanding is this stands in for a control output port that will be materialized, have a signal driven into it in this state, and hooked up to something in the datapath that is waiting for it (i.e. a calyx.group_go in this case).

If we are going to take the progressive lowering approach here, I think we need to better specify `fsm.ref`. Right now it feels like spooky action at a distance, that is specific to the current implementation.

mikeurbach replied:
Alternatively, if this is going to be a Calyx specific thing, why not just keep `calyx.enable` embedded within the FSM states, rather than defining this new thing, which feels Calyx specific.

mortbopet replied:
I missed this (fairly important) in the PR description: I would much rather prefer to embed `calyx.enable` (or more generally, arbitrary code within the `output` of a state). However, this requires a relaxation of the `calyx.enable` verifier, allowing it to be nested within something other than a Calyx control operation. It's a simple fix, however, it would make the Calyx dialect dependent on the FSM dialect, which might be controversial.

mortbopet replied:
addressed in the answer on the main thread.

mikeurbach reviewed (commented):
I haven't reviewed the implementation itself, or all the changes to FSM to make it happen, but I have some thoughts about this progressive lowering approach in general.

I think it might be possible to come up with good answers to my questions and make this approach work in general, and not just for this implementation.

However, I am also wondering if it wouldn't be a cleaner step to keep the FSM isolated from above, and just define the outputs it needs, rather than going through this `fsm.ref` mechanism, which feels specific to being embedded within Calyx.

mortbopet commented:
> However, I am also wondering if it wouldn't be a cleaner step to keep the FSM isolated from above, and just define the outputs it needs, rather than going through this fsm.ref mechanism, which feels specific to being embedded within Calyx.

I missed this (fairly important) in the PR description: I would much rather prefer to embed calyx.enable (or more generally, arbitrary code within the output of a state). However, this requires a relaxation of the calyx.enable verifier, allowing it to be nested within something other than a Calyx control operation. It's a simple fix, however, it would make the Calyx dialect dependent on the FSM dialect, which might be controversial.

As for whether creating the FSM and materializing the I/O in one go - it is possible, of course, but I'm hesitant to do such a solution if we see that there exists some intermediate design point (this PR).
This design point then needs to provide some value to be warranted: I think it's great in terms of readability/debugability of the generated FSM, and it *might* be simpler to transform the FSM while we only have `calyx.enable`s in the state outputs, instead of driving top-level I/O, but that's mostly based on a hunch.

cgyurgyik commented on lib/Dialect/Calyx/CalyxOps.cpp:
```diff
@@ -179,6 +179,12 @@ LogicalResult calyx::verifyCell(Operation *op) {
 LogicalResult calyx::verifyControlLikeOp(Operation *op) {
   auto parent = op->getParentOp();
+
+  if (!isa<calyx::CalyxDialect>(parent->getDialect())) {
```
Your comment says you relaxed the calyx enable verify, but this is relaxing all control-like operation verifiers?

mortbopet replied:
Correct; that might be a bit too relaxed, will update.

mortbopet commented:
Rebased onto main and dropped use of the `SymUniquer` in favor of `Namespace` - still interested in hearing people's thoughts on this, and whether it's something to move forward with - regardless, the FSMGraph changes should be merged.  
Locally I have an FSM materialization pass that takes the output of this pass, materializes I/O, and outlines the FSM - based on that, I can say that this representation is fairly easy to work with :slightly_smiling_face:.

mikeurbach commented:
> Locally I have an FSM materialization pass that takes the output of this pass, materializes I/O, and outlines the FSM - based on that, I can say that this representation is fairly easy to work with 🙂.

Can you share that, even if it's not polished?

The thing I'm still trying to work out is whether this partial lowering approach can be done in such a way where the embedded, partially lowered FSM is then materialized and outlined in a generic way, or if that is going to always be tied to Calyx. If it's possible to hide things behind interfaces, etc., then doing this partial lowering makes sense, and I think seeing the next couple steps after this PR will help understand that.

In general, using symbols lets you break out of explicitly passed SSA values, which may or may not be a good thing. I want to see what is gained by not passing the FSM inputs and outputs explicitly, because this symbol-based spooky action at a distance might be useful in some cases, but it does break locality, which comes with costs.

mikeurbach commented:
Another metric for success of this approach: can we do FSM-based optimizations in a way that doesn't have to care about Calyx? I.e. removing redundant states or something similar. If that can be done without caring about the fact that there are `calyx.enable`s inside the states, then great, we can say the FSM dialect takes an open world approach, and carefully design that.

My gut feeling is FSMs will be easier to reason about if they are isolated from above and take their inputs/outputs explicitly, but I'd love to see what is possible.

mortbopet commented:
Pushed the FSM materialization pass - for now, it just materializes the I/O and removes the `calyx.enable`s within the state output regions. Seeing as it's fairly small, merging the two passes into one - and in doing so, keeping `fsm.machine` `IsolatedFromAbove` - shouldn't be too cumbersome if we decide to go down that road.

> The thing I'm still trying to work out is whether this partial lowering approach can be done in such a way where the embedded, partially lowered FSM is then materialized and outlined in a generic way, or if that is going to always be tied to Calyx

Hmm... one possible interface could be an `ActivationInterface` (a bit like the `fsm.ref` op proposed earlier) with semantics:
* Activates some  Func-like symbol (can take N inputs and return M values)
* Activations to the same symbol are considered to materialize to a unique, shared interface
* Upon materialization, materializes to # cycles (to handle the static case) _or_ specifies some materialized output as a `done` signal.
* Upon materialization,extends top-level FSM I/O with:
  * M FSM inputs, or in the dynamic case, M + 1 (+`done` signal)
  * N + 1 FSM outputs (everything is expected to have some form of additional `go` signal).

That should be enough to handle both the latency sensitive and insensitive cases. Now, whether this is truly generic or just an excuse to hide `calyx.enable` behind an interface is up for debate.

> Another metric for success of this approach: can we do FSM-based optimizations in a way that doesn't have to care about Calyx?

If considering an open-world approach, i assume this would imply that whatever (potentially unregistered dialect) operation present within an output region must both be checked for uniqueness (is this already possible for `Operation*`s of unregistered dialect ops?) as well as declare whether they are side-effect free (to eliminate the spooky action at a distance). Again, coming up with an interface that fits our specific use-case here wouldn't be too hard, so it would be nice to have another client for this stuff to make sure that decisions are sane.

Thinking about it, getting around the latter will probably be a _massive_ pain point in writing generic optimizations for FSMs. Restricting FSM interactions with the outside world through the top-level interface sort of guarantees that the FSM at least has _some_ of knowledge of how each state is interacting with the outside world.

mikeurbach commented:
Thanks for sharing the materialization pass, and bearing with me in this discussion.

Taking a step back, I think your point about having another client for generic transformations before generalizing makes a lot of sense. Doing the Calyx-specific things in the short term definitely makes the most sense.

My question now is what do we sacrifice/gain to separate CalyxToFSM from MaterializeFSM, given that we are committing to not generalizing yet?

I see that we have to make generalize `calyx.enable` so it can be embedded in FSMs (or lower it to a fsm.ref construct, which would be Calyx-specific for now). We also need to make FSM no longer isolated from above.

This enables us to do the partial lowering, and separate the materialization to a later pass. I'm generally a big fan of the nanopass style, but here, I'm not sure it is worth it. Besides splitting the pass into smaller pieces, the checkpoint in the middle might be enable further transformation.

This is really the crux to me: are there meaningful (necessarily Calyx-specific) transformations you would want to do between CalyxToFSM and MaterializeFSM? Would it be better to do such transformations at this stage of partial lowering, or would it be better to do it after MaterializeFSM? If doing it after MaterializeFSM means it can be a generic transformation on the FSM dialect, rather than Calyx-specific, that seems like a reason to push to do such transformations after. 

My general feeling at this point is to combine MaterializeFSM into CalyxToFSM. At this point I think I've laid out my concerns and I can see the whole picture of what you're shooting for, so I'll leave it up to you.

Regardless of the approach, I will try to take a deeper look at the algorithm in CalyxToFSM.

rachitnigam commented:
> This is really the crux to me: are there meaningful (necessarily Calyx-specific) transformations you would want to do between CalyxToFSM and MaterializeFSM? 

Just to add one reason to separate the generation of FSMs from their materialization is to think about all the different FSMs that can be generated? For example, while working on the static-timing pass, we enumerated a couple of different FSM designs we could generate. Would it make sense to represent these different FSM designs as different materializeFSM passes?

rachitnigam commented:
Got it, thanks for the clarification. Apart from that, generalizing once a second use case shows up seems like a good idea. Loosening the calyx verifier for now to enable cleaner integration seems reasonable to me.

mikeurbach commented:
> leave all of the state transition logic to be materialized alongside the I/O.

I guess I had missed this part. This does seem to be the useful part of this intermediate state: simply moving symbolic things around before materializing the transitions and IO.

mortbopet commented:
Closing this PR in favor of starting a PR series.
